crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -160,18 +157,95 @@ impl SnapshotRepository {
 
         let upload = self
             .object_store
-            .put(&object_store::path::Path::from(snapshot_key), payload)
+            .put(
+                &object_store::path::Path::from(snapshot_key.clone()),
+                payload,
+            )
             .await
             .context("Failed to put snapshot in repository")?;
 
         debug!(
             %snapshot_id,
             etag = upload.e_tag.unwrap_or_default(),
             "Successfully published snapshot to repository as: {}",
-            key,
+            snapshot_key,
         );
         Ok(())
     }
+
+    pub(crate) async fn find_latest(
+        &self,
+        partition_id: PartitionId,
+    ) -> anyhow::Result<Option<LocalPartitionSnapshot>> {
+        let list_prefix = match self.prefix.as_str() {
+            "" | "/" => format!("{}/", partition_id),
+            prefix => format!("{}/{}/", prefix, partition_id),
+        };
+        let list_prefix = object_store::path::Path::from(list_prefix.as_str());
+
+        let list = self
+            .object_store
+            .list_with_delimiter(Some(&list_prefix))
+            .await?;
+
+        let latest = list.objects.first();
+
+        let Some(snapshot_entry) = latest else {
+            debug!(%partition_id, "No snapshots found in the snapshots repository");
+            return Ok(None);
+        };
+
+        let snapshot_object = self
+            .object_store
+            .get(&snapshot_entry.location)
+            .await
+            .context("Failed to get snapshot from repository")?;
+
+        // construct the bridge in a Tokio context, before moving to blocking pool
+        let snapshot_reader = SyncIoBridge::new(StreamReader::new(snapshot_object.into_stream()));
+
+        let snapshot_name = snapshot_entry.location.filename().expect("has a name");
+        let snapshot_base_path = &self.staging_path.join(snapshot_name);
+        tokio::fs::create_dir_all(snapshot_base_path).await?;
+
+        let snapshot_dir = snapshot_base_path.clone();
+        trace!(%partition_id, "Unpacking snapshot {} to: {:?}", snapshot_entry.location, snapshot_dir);
+        task_center()
+            .spawn_blocking_fn_unmanaged("unpack-snapshot", Some(partition_id), move || {
+                let mut tarball = tar::Archive::new(snapshot_reader);
+                for file in tarball.entries()? {
+                    let mut file = file?;
+                    trace!("Unpacking snapshot file: {:?}", file.header().path()?);
+                    file.unpack_in(&snapshot_dir)?;
+                }
+                Ok::<(), anyhow::Error>(())
+            })
+            .await??;
```

Comment:
I initially tried to use `spawn_blocking_unmanaged` here but got the following error:

> "Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks."

This is not a problem in the earlier use in this file - on line 136, we spawn a future which also uses the `tar` crate to bundle up a local directory on the upload path, and that works just fine. The only difference here appears to be the use of `SyncIoBridge` to make the `AsyncRead` byte stream of the download into a `Read` for tar's Archive to consume.

I was also able to get this to work using plain vanilla `tokio::task::spawn_blocking`. It seems like the TaskCenter implementation using `block_on` is part of this but I'm not entirely sure what's going on.

https://github.com/restatedev/restate/blob/c8257148955f73589bddf30b7bc87e3780bad3a0/crates/core/src/task_center.rs#L754-L756